File: nodejs/business-context/src/expression/logical/Negate.js
export default class Negate {
    constructor(arg) {
        this.arg = arg
    }

    interpret(context) {
        return !this.arg.interpret(context)
    }

    getName() {
        return 'logical-negate'
    }

    getArguments() {
        return [this.arg]
    }

    getProperties() {
        return {}
    }

    toString() {
        return 'not ' + this.arg.toString()
    }

    clone() {
        return new Negate(
            this.arg.clone()
        )
    }
}

File: java/business-context/src/main/java/cz/filipklimes/diploma/framework/businessContext/expression/logical/Negate.java
package cz.filipklimes.diploma.framework.businessContext.expression.logical;

import cz.filipklimes.diploma.framework.businessContext.expression.Expression;
import cz.filipklimes.diploma.framework.businessContext.expression.ExpressionVisitor;
import cz.filipklimes.diploma.framework.businessContext.expression.UnaryOperator;
import cz.filipklimes.diploma.framework.businessContext.weaver.BusinessOperationContext;

import java.util.*;

public class Negate extends UnaryOperator<Boolean, Boolean>
{

    public Negate(final Expression<Boolean> argument)
    {
        super(argument);
    }

    @Override
    public Boolean interpret(final BusinessOperationContext context)
    {
        return !getArgument().interpret(context);
    }

    @Override
    public Map<String, String> getProperties()
    {
        return Collections.emptyMap();
    }

    @Override
    public String getName()
    {
        return "logical-negate";
    }

    @Override
    public void accept(final ExpressionVisitor visitor)
    {
        visitor.visit(this);
    }

    @Override
    public String toString()
    {
        return String.format("not %s", getArgument());
    }

}
